docs: make Lua docstrings consistent

gpanders opened the pull request:
The official developer documentation in in `:h dev-lua-doc` specifies to use "--@" for special/magic tokens. However, this format is not consistent with EmmyLua notation (used by some Lua language servers) nor with the C version of the magic docstring tokens which use three comment characters.

Further, the code base is currently split between usage of "--@", "---@", and "--- @". In an effort to remain consistent, change all Lua magic tokens to use "---@" and update the developer documentation accordingly.

gpanders committed:
docs: make Lua docstrings consistent

The official developer documentation in in :h dev-lua-doc specifies to
use "--@" for special/magic tokens. However, this format is not
consistent with EmmyLua notation (used by some Lua language servers) nor
with the C version of the magic docstring tokens which use three comment
characters.

Further, the code base is currently split between usage of "--@",
"---@", and "--- @". In an effort to remain consistent, change all Lua
magic tokens to use "---@" and update the developer documentation
accordingly.

MunifTanjim commented:
The `---@param` format is used by Lua LSP server. They are EmmyLua annotation: https://emmylua.github.io/annotation.html

Are you sure you want to remove all these?

gpanders commented:
> The `---@param` format is used by Lua LSP server. They are EmmyLua annotation: https://emmylua.github.io/annotation.html
> 
> Are you sure you want to remove all these?

The Lua LSP server doesn't parse the `--@` style?

I don't have a personal opinion one way or another, but I do think it's important that we're a) consistent within the code base and b) matching whatever is documented. So if we need to keep the `---@` style then let's instead update the documentation to say that `--@` or `---@` should be used (although I think we should just use one or the other).

bfredl commented:
`---@param` is more similar to our C convention and IMO less surprising.

gpanders commented:
> `---@param` is more similar to our C convention and IMO less surprising.

Frankly I agree, I had always thought the `--@param` lines were a mistake until I read the mention in the developer docs earlier today.

I can amend the commit to prefer `---@` and also update the documentation accordingly.

MunifTanjim commented:
> The Lua LSP server doesn't parse the --@ style?

Nope.

gpanders commented:
@justinmk Seeing as you wrote the docs on Lua docstrings in `develop.txt`, would you like to weigh in? Any reason you chose `--@` over `---@`?

gpanders commented:
The [`lua2dox.lua` script supports both `--@` and `---@`](https://github.com/neovim/neovim/blob/master/scripts/lua2dox.lua#L409-L416) so this should be fine.

vigoux reviewed (commented):
There's a bunch of change to the help files that I think don't belong here. Is this a rebase/merge error of some sort ?

gpanders commented:
> There's a bunch of change to the help files that I think don't belong here. Is this a rebase/merge error of some sort ?

The changes you're seeing are intentional:

1. `restore_extmarks` and `save_extmarks` were marked as `@private` and should never have been in the docs in the first place. Changing `--- @private` to `---@private` allowed the docs generator to correctly pick this up.
2. Some methods, such as `get_lines` and `rename`, used only `--` for their public docstring comment. I changed these to `---` so that they would be included in the docs.
3. [`vim.notify` had a two-line docstring but without a newline in between them](https://github.com/neovim/neovim/pull/15255/files#diff-a8fd4e44d96101de6e4453a16811a686ce91e33e4767af7666481edb338d0744L410-L416) which caused the resulting docstring in the help file to be merged into one line. I added a newline between these two lines to ensure the output also had a newline.

I think that's all of them.

vigoux commented:
There's a conflict with this PR, could you rebase on latest master ?

clason commented:
(also, commit type should be `docs` -- `chore` is a last resort if none of the other types is applicable)

vigoux commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -577,28 +577,28 @@ end
 --- Sorts by CompletionItem.sortText.
 ---
---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion
+--see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion
```
Shouldn't this be 
```suggestion
---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion
```

gpanders replied:
I can update these, but since they weren't already included in the docs I chose to err toward leaving it how it is.

vigoux commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -577,28 +577,28 @@ end
 --- Returns text that should be inserted when selecting completion item. The
 --- precedence is as follows: textEdit.newText > insertText > label
---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion
+--see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion
```
Same remark as above ?

vigoux commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -754,8 +754,8 @@ end
---@param workspace_edit (table) `WorkspaceEdit`
--- @see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_applyEdit
+---@param workspace_edit (table) `WorkspaceEdit`
+--see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_applyEdit
```
```suggestion
---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_applyEdit
```

vigoux reviewed (changes requested):
Just a littel question: why using `---@` instead of `--- @` (with a space) ?

In our C codebase, the style is generally `/// @` and not `///@`.

gpanders commented:
> Just a littel question: why using `---@` instead of `--- @` (with a space) ?
> 
> In our C codebase, the style is generally `/// @` and not `///@`.

I agree this looks better and is more consistent, but alas both the [Lua docs generator](https://github.com/neovim/neovim/blob/master/scripts/lua2dox.lua#L414) and the [EmmyLua annotations](https://emmylua.github.io/annotations/class.html) (which are used by Lua LSPs) use `---@` without a space.

vigoux commented:
> > Just a littel question: why using `---@` instead of `--- @` (with a space) ?
> > In our C codebase, the style is generally `/// @` and not `///@`.
> 
> I agree this looks better and is more consistent, but alas both the [Lua docs generator](https://github.com/neovim/neovim/blob/master/scripts/lua2dox.lua#L414) and the [EmmyLua annotations](https://emmylua.github.io/annotations/class.html) (which are used by Lua LSPs) use `---@` without a space.

That's a weird design decision from them then, but if that's the way it works, then let's do it.

justinmk commented:
> @justinmk Seeing as you wrote the docs on Lua docstrings in `develop.txt`, would you like to weigh in? Any reason you chose `--@` over `---@`?

No reason, it was the only thing our hacky docs generator supported at the time. Thank you for this PR!

justinmk committed:
Merge branch 'master' into lua-docs
